### PR TITLE
[synthetic-data] Add on-the-fly synthetic length-distribution dataset

### PR DESCRIPTION
**Pull request by jinsooihm**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #4275

Streaming data source for DP load-balancing and throughput experiments,
generating random-token sequences whose lengths follow a configurable
distribution. No dataset is written to disk.

Example usage:

```python
from torchtitan.components.data import (
    BucketLengthSpec,
    FirstFitPackingConfig,
    GrainDataLoader,
    LengthBucket,
    RandomTokenProcessor,
    SingleDatasetConfig,
    SyntheticLengthSource,
)

synthetic_ds = SingleDatasetConfig(
    source=SyntheticLengthSource.Config(
        length_spec=BucketLengthSpec(
            buckets=(
                # keep max_len < training.max_context_length; longer draws are dropped
                LengthBucket(min_len=1, max_len=128, weight=3.0),
                LengthBucket(min_len=1024, max_len=2000, weight=1.0),
            ),
        ),
        seed=0,
    ),
    processor=RandomTokenProcessor.Config(),  # vocab from tokenizer
    post_filters=(lambda sample: sample is not None,),
)

# Pack whole sequences into rows, as SFT does. FirstFit preserves each sampled
# length; ConcatThenSplit would re-chunk into uniform blocks and erase the
# distribution you configured.
synthetic_packed_ds = FirstFitPackingConfig(dataset=synthetic_ds)

config.dataloader = GrainDataLoader.Config(dataset=synthetic_packed_ds, shuffle=False)
```
Another option is to use `ConstantTokenProcessor` producing same token for perhaps non-MoE models in which token routing does not matter.

`scripts/preview_synthetic_lengths.py`: inspect a distribution (percentiles, histogram, DP-balance simulation) for a seed without running training.

Tests: tests/unit_tests/components/data/test_synthetic_data.py and
tests/unit_tests/test_preview_synthetic_lengths.py (30 tests).

### CODE REVIEW

File: torchtitan/components/data/README.md [hunk 1]
```diff
@@ -175,6 +175,69 @@ packed_tokens_ds = ConcatThenSplitPackingConfig(
...
+# Pack whole sequences into rows, as SFT does. FirstFit preserves each sampled
+# length; ConcatThenSplit would re-chunk into uniform blocks and erase the
+# distribution you configured.
+synthetic_packed_ds = FirstFitPackingConfig(dataset=synthetic_ds)
+
+config.dataloader = GrainDataLoader.Config(dataset=synthetic_packed_ds, shuffle=False)
+```
+
+Notes:
+- Content is random and irrelevant for perf/load-balancing; only the length
```
Comment (tianyu-l) [inline torchtitan/components/data/README.md]: For MoE load balance, it is

File: torchtitan/components/data/README.md [hunk 2]
```diff
@@ -175,6 +175,69 @@ packed_tokens_ds = ConcatThenSplitPackingConfig(
 )
 ```
 
+## Synthetic length-distribution data (experiments)
+
+For DP load-balancing and throughput experiments you often want sequences whose
```
Comment (tianyu-l) [inline torchtitan/components/data/README.md]: Do we run this E2E? What if loss / grad norm explode?

File: torchtitan/components/data/README.md (hunk 1 above)
Comment (jinsooihm) [reply torchtitan/components/data/README.md]: Noob question, but is testing MoE on random tokens a fair way to test if token routing matters?

File: torchtitan/components/data/README.md (hunk 2 above)
Comment (jinsooihm) [reply torchtitan/components/data/README.md]: Hmm, my guess would be that if the tokens are random, then the loss / grad would converge to whatever random prediction gives, which should be a finite value?

Comment (tianyu-l) [reply torchtitan/components/data/README.md]: I think loss would be. Grad norm, I'm not sure

File: torchtitan/components/data/README.md (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/components/data/README.md]: No, I'm just saying that the note is not accurate

Comment (jinsooihm) [reply torchtitan/components/data/README.md]: Yes, I understand that. I am just curious about what kind of dataset one should use when doing perf comparisons on MoE.

File: torchtitan/components/data/README.md (hunk 2 above)
Comment (jinsooihm) [reply torchtitan/components/data/README.md]: This is what claude tells:

The gradient of cross-entropy w.r.t. the logits is always

$$g = \text{softmax}(z) - e_y = p - e_y.$$

Both $p$ and $e_y$ are probability vectors (entries in [0,1] summing to 1), so their difference is bounded no matter what:

$$|g|_1 \le |p|_1 + |e_y|_1 = 2.$$

That bound is independent of the token/label and of vocab size — a random label gives the same $p - e_y$ form as a real one, so synthetic data can't make it bigger.

Everything downstream is just this bounded signal times the network's Jacobian. Normalization (RMSNorm) caps activation sizes and residual connections keep each layer's Jacobian ~O(1), so a bounded input stays bounded through backprop. And clip_grad_norm_(max_norm=1.0) caps the applied update at 1.0 regardless.

And test running llama3b debugmodel for 1000 steps:
Screenshot 2026-08-21 at 5 14 34 PM

File: torchtitan/components/data/README.md (hunk 1 above)
Comment (tianyu-l) [reply torchtitan/components/data/README.md]: must be real dataset

File: torchtitan/components/data/README.md [hunk 3]
```diff
@@ -175,6 +175,69 @@ packed_tokens_ds = ConcatThenSplitPackingConfig(
 )
 ```
 
+## Synthetic length-distribution data (experiments)
```
Comment (felipemello1) [inline torchtitan/components/data/README.md]: I feel like this is an edge case and should be at the bottom of the readme, not in the middle of it. I also think its a bit too long as it to be in the readme. We could add a small paragraph and point to the docstring of the class, which would hold al the details, and/or have a dataloader predefined somewhere that people can import, e.g.

```python
from data.synthetic import dataloader_builder # this is a bad name, lets follow the convention used in other files
```

```python
def synthetic_dataloader_builder()
	synthetic_packed_ds = ...
	synthetic_dataloader = GrainDataLoader.Config(dataset=synthetic_packed_ds, shuffle=False)
	return synthetic_dataloader
```

File: torchtitan/components/data/synthetic.py [hunk 4]
```diff
@@ -0,0 +1,250 @@
...
+
+    def __call__(self, sample: dict, rng: np.random.Generator) -> TextSequence | None:
+        length = int(sample["length"])
+        if _should_drop(length, self._max_context_length):
+            return None
+        ids = rng.integers(0, self._vocab_size, size=length + 1, dtype=np.int64)
+        return _split_next_token(ids)
+
+
+class ConstantTokenProcessor(SampleProcessor):
```
Comment (felipemello1) [inline torchtitan/components/data/synthetic.py]: is there a strong reason to keep the `Constant` version, or can we just keep the `Random` one?

File: torchtitan/components/data/synthetic.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+            raise ValueError("BucketLengthSpec requires at least one bucket")
+
+    def sample(self, rng: np.random.Generator, size: int) -> np.ndarray:
+        weights = np.array([b.weight for b in self.buckets], dtype=np.float64)
+        probs = weights / weights.sum()
+        choices = rng.choice(len(self.buckets), size=size, p=probs)
+        lows = np.array([b.min_len for b in self.buckets], dtype=np.int64)
+        highs = np.array([b.max_len for b in self.buckets], dtype=np.int64)
+        lengths = rng.integers(lows[choices], highs[choices] + 1)
+        return lengths.astype(np.int64)
```
Comment (felipemello1) [inline torchtitan/components/data/synthetic.py:83]: This .sample() i called on every new sample produced IIUC. Most of these arrays should be created only once, not on every sample. Perhaps the optimization doesnt matter too much here, but it seems like a small lift.

File: torchtitan/components/data/synthetic.py [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+
+
+@dataclass(frozen=True, kw_only=True, slots=True)
+class ParametricLengthSpec:
```
Comment (felipemello1) [inline torchtitan/components/data/synthetic.py]: Do we need 2 types of lengthspec or `BucketLengthSpec` is good enough?

File: torchtitan/components/data/synthetic.py [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+
+
+class _SyntheticLengthIterator(grain.DatasetIterator):
+    """Exposes the length-sampling RNG to Grain checkpoint recursion.
+
+    Lengths are drawn in chunks so per-record NumPy dispatch does not distort
+    throughput/DP-balance measurements.
+    """
+
+    _CHUNK = 1024
```
Comment (felipemello1) [inline torchtitan/components/data/synthetic.py]: in "BucketLengthSpec" you call this `size`. Let's use the same naming conventions everywhere.

File: torchtitan/components/data/synthetic.py [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+
+    def __iter__(self) -> grain.DatasetIterator:
+        return _SyntheticLengthIterator(self._length_spec, self._seed)
+
+
+class _SyntheticLengthIterator(grain.DatasetIterator):
+    """Exposes the length-sampling RNG to Grain checkpoint recursion.
+
+    Lengths are drawn in chunks so per-record NumPy dispatch does not distort
+    throughput/DP-balance measurements.
```
Comment (felipemello1) [inline torchtitan/components/data/synthetic.py]: i dont follow this. What is the difference between sampling 1024 at once vs 1 at a time?

File: torchtitan/components/data/synthetic.py (hunk 5 above)
Comment (felipemello1) [reply torchtitan/components/data/synthetic.py:83]: edit: i see that you have a buffer. I guess its not a big deal.

File: torchtitan/components/data/synthetic.py [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+        }
+
+    def set_state(self, state: dict) -> None:
+        self._rng.bit_generator.state = state["bit_generator"]
+        self._buffer = np.array(state["remaining"], dtype=np.int64)
+        self._offset = 0
+
+
+def _should_drop(length: int, max_context_length: int) -> bool:
+    return length < 1 or length + 1 > max_context_length
```
Comment (felipemello1) [inline torchtitan/components/data/synthetic.py]: Would this ever happen if we are sampling from a bucket with size defined between [1, max_context_length]? If the answer is no, lets delete and simplify.

File: torchtitan/components/data/synthetic.py [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+
+
+def _split_next_token(ids: np.ndarray) -> TextSequence:
+    return TextSequence(input_ids=ids[:-1], labels=ids[1:])
```
Comment (felipemello1) [inline torchtitan/components/data/synthetic.py]: i dont think that "split" is the best word here. Can you check how we name it in other datasets and use the same pattern?

File: scripts/preview_synthetic_lengths.py
Comment (felipemello1) [inline scripts/preview_synthetic_lengths.py:1]: I think that this script is not mentioned in the readme of the synthetic.py. Its probably worth a mention in the synthetic.py file, at the top.

**Review by felipemello1 — COMMENTED** [review]
Thanks for the PR! I did a first pass. My 2c is that we can probably:

a) delete/simplify some code (examples in the comment)
b) Move the bulk of readme changes to docstring at the top of synthetic.py
c) Have a dataloader builder in synthetic.py that users can just import to use, rather then showing in the readme how to do it

I didn't rerun the PR, but i am trusting you checked the outputs, ran training, and the generation looked like what you expected.

File: torchtitan/components/data/synthetic.py (hunk 6 above)
Comment (jinsooihm) [reply torchtitan/components/data/synthetic.py]: Hmm, I wasn't sure what people would need, but probably BucketLengthSpec is enough?

File: torchtitan/components/data/synthetic.py (hunk 8 above)
Comment (jinsooihm) [reply torchtitan/components/data/synthetic.py]: Sampling 1024 at once makes less Python to C dispatch, so it is more efficient.

File: torchtitan/components/data/synthetic.py (hunk 10 above)
Comment (jinsooihm) [reply torchtitan/components/data/synthetic.py]: It just seems to be inlined else where, like in `hf_datasets/text_datasets.py`. I removed the function.

File: torchtitan/components/data/synthetic.py (hunk 9 above)
Comment (jinsooihm) [reply torchtitan/components/data/synthetic.py]: It shouldn't if correctly configured. I removed it.

File: torchtitan/components/data/synthetic.py [hunk 11]
```diff
@@ -0,0 +1,250 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""On-the-fly synthetic token sequences with a configurable length distribution.
+
+Intended for DP load-balancing and throughput experiments where token *content*
+is irrelevant and only the sequence-length distribution matters.
```
Comment (fegin) [inline torchtitan/components/data/synthetic.py]: You can mention that synthetic dataset should be use with MoE force balanced option, to ensure that MoE's performance not affected by the random content.

File: torchtitan/components/data/synthetic.py (hunk 4 above)
Comment (fegin) [reply torchtitan/components/data/synthetic.py]: `RandomTokenProcessor` can also be used by non-MoE model without affecting anything. So I suggest we just remove `Constant` one and inline `_split_next_token` to `RandomTokenProcessor`.

File: torchtitan/components/data/synthetic.py (hunk 11 above)
Comment (fegin) [inline torchtitan/components/data/synthetic.py]: Probably also mention that if users can use force balanced option in MoE if they don't want the MoE routing affected by the random dataset.

File: tests/unit_tests/components/data/test_synthetic_data.py [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+def _loader():
+    return GrainDataLoader.Config(
+        dataset=SingleDatasetConfig(
+            source=SyntheticLengthSource.Config(
+                length_spec=BucketLengthSpec(
+                    buckets=(LengthBucket(min_len=2, max_len=6),)
+                ),
+                seed=0,
+            ),
+            processor=RandomTokenProcessor.Config(vocab_size=100),
```
Comment (fegin) [inline tests/unit_tests/components/data/test_synthetic_data.py]: We actually generate the dataset in RandomTokenProcessor, not in the source. I was imagining the source should be the dataset.  Any reason to decouple them instead of just one `SyntheticDataset` as the source?

Comment (jinsooihm) [reply tests/unit_tests/components/data/test_synthetic_data.py]: I was thinking that the source produces length as data and `RandomTokenProcessor` is tokenizing the data to produce the actual sequence. But since we are removing `ConstantTokenProcessor`, we can remove this decoupling.
